### PR TITLE
python3Packages.qiskit: 0.19.6 -> 0.20.0 & update components

### PR DESCRIPTION
**Pull request by drewrisinger**
###### Motivation for this change
Updates ``python3Packages.qiskit`` to v0.20.0, as well as updates all components.
Some components require new packages, which were added (``muparserx, pythonPackages.yfinance (which requires pythonPackages.multitasking)``).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/muparserx/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{ stdenv
+, fetchFromGitHub
+, cmake
+  # Check Inputs
+, python
```
Comment (jonringer) [inline pkgs/development/libraries/muparserx/default.nix:5]: unused
```suggestion
```

File: pkgs/development/python-modules/retworkx/default.nix [hunk 2]
```diff
@@ -1,62 +1,65 @@
 { lib
-, buildPythonPackage
-, pythonOlder
-, pythonAtLeast
-, python
+, rustPlatform
+, python3
```
Comment (jonringer) [inline pkgs/development/python-modules/retworkx/default.nix:3]: inside of a python package set, python will point to the correct interpreter version. python3 is likely to only be aliased to python38.
Causing issues for the python37 version
```suggestion
, python
```

File: pkgs/development/python-modules/retworkx/default.nix [hunk 3]
```diff
@@ -1,62 +1,65 @@
...
+      } )
+  ];
 
-  src = fetchPypi {
-    inherit pname version format;
-    inherit (wheel-args) python sha256;
-    abi = if pythonOlder "3.8" then "${wheel-args.python}m" else wheel-args.python;
-    platform = "manylinux2010_x86_64"; # i686, aarch64, and ppc64 also available, restricting to x86 for simplicity
-  };
+  propagatedBuildInputs = [ python3 ];
```
Comment (jonringer) [inline pkgs/development/python-modules/retworkx/default.nix:33]:
```suggestion
  propagatedBuildInputs = [ python ];
```

File: pkgs/development/python-modules/yfinance/default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  propagatedBuildInputs = [
+    multitasking
+    numpy
+    pandas
+    requests
+  ];
+
+  doCheck = false;  # Tests require internet access
+  pythonImportsCheck = ["yfinance"];
```
Comment (jonringer) [inline pkgs/development/python-modules/yfinance/default.nix:28]:
```suggestion
  pythonImportsCheck = [ "yfinance" ];
```

**Review by jonringer — COMMENTED** [review]
ohterwise LGTM

File: pkgs/development/python-modules/retworkx/default.nix [hunk 5]
```diff
@@ -1,62 +1,65 @@
...
+
+  nativeBuildInputs = [ pipInstallHook maturin ];
```
Comment (jonringer) [inline pkgs/development/python-modules/retworkx/default.nix:35]:
```suggestion
  nativeBuildInputs = [ pipInstallHook maturin  pip ];
```
```
error: build of '/nix/store/lchz92l48c5spy3arjh3grwdjal364ay-retworkx-0.4.0.drv' on 'ssh://build' failed: builder for '/nix/store/lchz92l48c5spy3arjh3grwdjal364ay-retworkx-0.4.0.drv' failed with exit code 1
builder for '/nix/store/lchz92l48c5spy3arjh3grwdjal364ay-retworkx-0.4.0.drv' failed with exit code 1; last 10 log lines:
     Compiling indoc v0.3.6
     Compiling numpy v0.11.0
     Compiling retworkx v0.4.0 (/build/source)
      Finished release [optimized] target(s) in 21.20s
  📦 Built wheel for CPython 3.7m to /build/source/target/wheels/retworkx-0.4.0-cp37-cp37m-linux_x86_64.whl
  installing
  Executing pipInstallPhase
  /build/source/dist /build/source
  /nix/store/9hdzs3g6rb76grwzlsbxyxwxg2n0fl4a-python3-3.7.8/bin/python3.7: No module named pip
  builder for '/nix/store/lchz92l48c5spy3arjh3grwdjal364ay-retworkx-0.4.0.drv' failed with exit code 1
```

**Review by jonringer — COMMENTED** [review]
also, retworkx in `python-modules` needs to be changed since it's not a normal `buildPythonPackage`.

original:
```
  retworkx = callPackage ../development/python-modules/retworkx { };
```
correct:
```
  retworkx = toPythonModule(callPackage ../development/python-modules/retworkx { });
```

**Comment by jonringer** [comment]
do you mind looking at https://github.com/NixOS/nixpkgs/issues/95527 as well?

File: pkgs/development/libraries/muparserx/default.nix (hunk 1 above)
Comment (drewrisinger) [reply pkgs/development/libraries/muparserx/default.nix:5]: TY. This made me remember I wanted to actually build & run its self-test module. I'll try to do that.

**Comment by drewrisinger** [comment]
Addressed all comments. Building locally now. Only outstanding item IMHO is maybe running self-test on ``muparserx`` (WIP).

🤦 on changing ``retworkx = toPythonModule ...``, did it in my ``nur-packages`` but forgot here... Fixed.

**Comment by drewrisinger** [comment]
OK. Worked out last few bugs. 
* ``pythonPackages.retworkx`` now properly works as ``python37Packages.retworkx``, and is disabled for ``python27Packages``.
* muparserx is self-tested
* all review comments addressed.
* ``nix-review pr 95713`` runs locally

**Review by jonringer — COMMENTED** [review]
seeing some issues with aqua:
```
https://github.com/NixOS/nixpkgs/pull/95713
6 packages failed to build:
python37Packages.qasm2image python37Packages.qiskit python37Packages.qiskit-aqua python38Packages.qasm2image python38Packages.qiskit python38Packages.qiskit-aqua

20 packages built:
muparserx python27Packages.multitasking python27Packages.pylatexenc python27Packages.yfinance python37Packages.multitasking python37Packages.pylatexenc python37Packages.qiskit-aer python37Packages.qiskit-ibmq-provider python37Packages.qiskit-ignis python37Packages.qiskit-terra python37Packages.retworkx python37Packages.yfinance python38Packages.multitasking python38Packages.pylatexenc python38Packages.qiskit-aer python38Packages.qiskit-ibmq-provider python38Packages.qiskit-ignis python38Packages.qiskit-terra python38Packages.retworkx python38Packages.yfinance
```
the failure:
```
=================================== FAILURES ===================================
__________________________ TestOptimizers.test_p_bfgs __________________________

self = <test.aqua.test_optimizers.TestOptimizers testMethod=test_p_bfgs>

    def test_p_bfgs(self):
        """ parallel l_bfgs_b test """
        optimizer = P_BFGS(maxfun=1000)
        res = self._optimize(optimizer)
>       self.assertLessEqual(res[2], 10000)
E       AssertionError: 44250 not less than or equal to 10000

test/aqua/test_optimizers.py:69: AssertionError
```

**Comment by drewrisinger** [comment]
> 
> the failure:
> 
> ```
> =================================== FAILURES ===================================
> __________________________ TestOptimizers.test_p_bfgs __________________________
> 
> self = <test.aqua.test_optimizers.TestOptimizers testMethod=test_p_bfgs>
> 
>     def test_p_bfgs(self):
>         """ parallel l_bfgs_b test """
>         optimizer = P_BFGS(maxfun=1000)
>         res = self._optimize(optimizer)
> >       self.assertLessEqual(res[2], 10000)
> E       AssertionError: 44250 not less than or equal to 10000
> 
> test/aqua/test_optimizers.py:69: AssertionError
> ```
@jonringer this is probably another massively-parallel CPU issue... the test runs an optimizer that distributes across the cores (https://github.com/Qiskit/qiskit-aqua/blob/72a6fb3b40accfe18b77bfb8c811aa94a7afa443/qiskit/aqua/components/optimizers/p_bfgs.py#L90), which each run an optimization and the number of function evalutions are summed across the processes and assumed to be < 10000 in the test.

I'll open an upstream issue, but in the meantime I feel comfortable disabling this test to get this PR to pass since I've not had this issue locally or on the GitHub Actions build of my NUR repo.

**Comment by jonringer** [comment]
```
Steps to reproduce the problem

    Get a Threadripper 3990X (or equivalent many-core CPU), running Linux
```
lol, yes, $4k cpu :)

**Comment by drewrisinger** [comment]
> ```
> Steps to reproduce the problem
> 
>     Get a Threadripper 3990X (or equivalent many-core CPU), running Linux
> ```
> 
> lol, yes, $4k cpu :)

That is the CPU you tested it against, right? 
And this package is distributed by IBM, I'm sure they have a few high-core-count systems like Summit that they could test this on if they really wanted :D

**Comment by jonringer** [comment]
right, or the 7742 also has 64/128 https://www.newegg.com/amd-epyc-7742-socket-sp3/p/N82E16819113581

**Comment by drewrisinger** [comment]
@jonringer pushed the patch upstream made. Hoping ``python3Packages.qiskit-aqua`` will build cleanly for you now.

**Review by jonringer — APPROVED** [review]
diff LGTM

```
https://github.com/NixOS/nixpkgs/pull/95713
26 packages built:
muparserx python27Packages.multitasking python27Packages.pylatexenc python27Packages.yfinance python37Packages.multitasking python37Packages.pylatexenc python37Packages.qasm2image python37Packages.qiskit python37Packages.qiskit-aer python37Packages.qiskit-aqua python37Packages.qiskit-ibmq-provider python37Packages.qiskit-ignis python37Packages.qiskit-terra python37Packages.retworkx python37Packages.yfinance python38Packages.multitasking python38Packages.pylatexenc python38Packages.qasm2image python38Packages.qiskit python38Packages.qiskit-aer python38Packages.qiskit-aqua python38Packages.qiskit-ibmq-provider python38Packages.qiskit-ignis python38Packages.qiskit-terra python38Packages.retworkx python38Packages.yfinance
```

cc @FRidh does the rust -> python `retworkx` changes look good to you?

**Comment by jonringer** [comment]
> @jonringer pushed the patch upstream made. Hoping `python3Packages.qiskit-aqua` will build cleanly for you now.

I think i mentioned it somewhere else, but i had to drop the `NIX_BUILD_CORES` to 32 due to other packages having timing issues (bashCompeletion would fail in tests, some meson builds would try to link against objects that don't exist, etc)

**Comment by drewrisinger** [comment]
> cc @FRidh does the rust -> python `retworkx` changes look good to you?

I based it on the @FRidh approved https://github.com/NixOS/nixpkgs/pull/91342 ``python3Packages.tokenizers``.

**Pull request merged**